packages/dataviews/src/components/dataform/index.tsx

```diff
@@ -76,6 +77,29 @@ function DataFormNumberControl< Item >( {
 		[ id, onChange ]
 	);
 
+	if ( field.elements ) {
```

Comment:
I'm personally wondering whether this should be another type or not `enum`. I'm curious if @joshuatf has an opinion.

My feeling is that we should have a way to define these "controls" and pick a control by assigning a "type" or something else. I'm not sure that I like the fact that other field properties can be taken into consideration to pick the actual control.

Reply:
To expand a bit. having a unique discriminator would be good for field "types" or "controls" registration.

Reply:
The `type` is used for validation, so it needs to know the actual type (string, integer, etc.) of the underlying data. If we want the consumer to decide exactly which control to use we'd need a different prop field. This can be a follow-up PR.

Reply:
> The type is used for validation, so it needs to know the actual type (string, integer, etc.) of the underlying data. 

For me this is another argument for `type: "enum"` as the validation for this particular case is to actually check whether the value is within the "elements" array rather than checking whether it's integer or string.

Reply:
> To expand a bit. having a unique discriminator would be good for field "types" or "controls" registration.

Personally, I think this is the point that I care about the most. If you think we should keep type: "integer" here that works for me but I would suggest changing the implementation and make sure the `IntegerControl` switches between `Number` and `Select` depending on whether there's a `elements` or not (and same for `TextControl`).

Reply:
We actually had the `enumeration` type but removed it a few months ago after some discussions https://github.com/WordPress/gutenberg/pull/60084 I think those concerns are still valid. Introducing field types is stress-testing our APIs (see example about [sorting](https://github.com/WordPress/gutenberg/pull/63983#discussion_r1692936050)), and I'd like to formalize them a bit more using the underlying data type (same as REST API types: string, number, etc.).

> as the validation for this particular case is to actually check whether the value is within the "elements" array rather than checking whether it's integer or string

This is a good point! I's both to me: it should be integer AND it should be part of the provided integer list. I'll update to consider this as well.

Reply:
> Personally, I think this is the point that I care about the most. If you think we should keep type: "integer" here that works for me but I would suggest changing the implementation and make sure the IntegerControl switches between Number and Select depending on whether there's a elements or not (and same for TextControl). I know it means some duplication but for me it's still better.

That's how it is implemented now. With the caveat that I didn't implemented for the text type because it's not necessary so far. To keep the PRs minimal I thought it best to do it when it was need it, but I can also add it here if that's preferred. A PR should be up soon for this anyway: we need it for quick editing status.

Reply:
> That's how it is implemented now. With the caveat that I didn't implemented for the text type because it's not necessary so far. 

I missed that sorry :P I thought the check was at the parent level. All good for me for now then.